vault/rollback.go

```diff
@@ -184,17 +193,39 @@ func (m *RollbackManager) attemptRollback(ctx context.Context, fullPath string,
 		Path:      ns.TrimmedPath(fullPath),
 	}
 
+	releaseLock := true
 	if grabStatelock {
+		doneCh := make(chan struct{})
+		defer close(doneCh)
+
+		stopCh := make(chan struct{})
+		go func() {
+			defer close(stopCh)
+
+			select {
+			case <-m.shutdownCh:
+			case <-rs.cancelLockGrabCtx.Done():
+			case <-doneCh:
+			}
+		}()
+
 		// Grab the statelock or stop
-		if stopped := grabLockOrStop(m.core.stateLock.RLock, m.core.stateLock.RUnlock, m.shutdownCh); stopped {
-			return errors.New("rollback shutting down")
+		if stopped := grabLockOrStop(m.core.stateLock.RLock, m.core.stateLock.RUnlock, stopCh); stopped {
+			// If we stopped due to shutdown, return. Otherwise another thread
+			// is holding the lock for us, continue on.
+			select {
+			case <-m.shutdownCh:
```

Comment:
Will it be possible to read from this channel both in here and in the goroutine above?

Reply:
In the above goroutine, in case of a shutdown, stopCh will be closed by reading `m.shutdownCh`. We are trying to read from the same again here.

Reply:
You can read from a closed channel multiple times

Reply:
Remember that a receive from a closed channel returns immediately.  So that's okay, we're not trying to consume anything, just use it as a signal.